S390x ci periodic tests

AlekseiNikiforovIBM opened the pull request:
Periodically run testsuite for s390x

**Dependencies update**
Package z3-solver is updated from version 4.12.2.0 to version 4.12.6.0. This is a minor version update, so no functional change is expected.
The reason for update is build on s390x. pypi doesn't provide binary build for z3-solver for versions 4.12.2.0 or 4.12.6.0 for s390x. Unfortunately, version 4.12.2.0 fails to build with newer gcc used on s390x builders, but those errors are fixed in version 4.12.6.0. Due to this minor version bump fixes build on s390x.

```
# pip3 install z3-solver==4.12.2.0
...
      In file included from /tmp/pip-install-756iytc6/z3-solver_ce6f750b780b4146a9a7c01e52672071/core/src/util/region.cpp:53:
      /tmp/pip-install-756iytc6/z3-solver_ce6f750b780b4146a9a7c01e52672071/core/src/util/region.cpp: In member function ‘void* region::allocate(size_t)’:
      /tmp/pip-install-756iytc6/z3-solver_ce6f750b780b4146a9a7c01e52672071/core/src/util/tptr.h:29:62: error: ‘uintptr_t’ does not name a type
         29 | #define ALIGN(T, PTR) reinterpret_cast<T>(((reinterpret_cast<uintptr_t>(PTR) >> PTR_ALIGNMENT) + \
            |                                                              ^~~~~~~~~
      /tmp/pip-install-756iytc6/z3-solver_ce6f750b780b4146a9a7c01e52672071/core/src/util/region.cpp:82:22: note: in expansion of macro ‘ALIGN’
         82 |         m_curr_ptr = ALIGN(char *, new_curr_ptr);
            |                      ^~~~~
      /tmp/pip-install-756iytc6/z3-solver_ce6f750b780b4146a9a7c01e52672071/core/src/util/region.cpp:57:1: note: ‘uintptr_t’ is defined in header ‘<cstdint>’; did you forget to ‘#include <cstdint>’?
         56 | #include "util/page.h"
        +++ |+#include <cstdint>
         57 |
```

**Python paths update**
On AlmaLinux 8 s390x, old paths:
```
python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())'
/usr/lib/python3.12/site-packages
```

Total result is `/usr/lib/python3.12/site-packages/torch;/usr/lib/python3.12/site-packages`

New paths:
```
python -c 'import site; print(";".join([x for x in site.getsitepackages()] + [x + "/torch" for x in site.getsitepackages()]))'
/usr/local/lib64/python3.12/site-packages;/usr/local/lib/python3.12/site-packages;/usr/lib64/python3.12/site-packages;/usr/lib/python3.12/site-packages;/usr/local/lib64/python3.12/site-packages/torch;/usr/local/lib/python3.12/site-packages/torch;/usr/lib64/python3.12/site-packages/torch;/usr/lib/python3.12/site-packages/torch
```

```
# python -c 'import torch ; print(torch)'
<module 'torch' from '/usr/local/lib64/python3.12/site-packages/torch/__init__.py'>
```

`pip3 install dist/*.whl` installs torch into `/usr/local/lib64/python3.12/site-packages`, and later it's not found by cmake with old paths:

```
CMake Error at CMakeLists.txt:9 (find_package):
  By not providing "FindTorch.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Torch", but
  CMake did not find one.
```

https://github.com/pytorch/pytorch/actions/runs/10994060107/job/30521868178?pr=125401

**Builders availability**
Build took 60 minutes
Tests took: 150, 110, 65, 55, 115, 85, 50, 70, 105, 110 minutes (split into 10 shards)

60 + 150 + 110 + 65 + 55 + 115 + 85 + 50 + 70 + 105 + 110 = 975 minutes used. Let's double it. It would be 1950 minutes.

We have 20 machines * 24 hours = 20 * 24 * 60 = 20 * 1440 = 28800 minutes

We currently run 5 nightly binaries builds, each on average 90 minutes build, 15 minutes test, 5 minutes upload, 110 minutes total for each, 550 minutes total. Doubling would be 1100 minutes.

That leaves 28800 - 1100 = 27700 minutes total. Periodic tests would use will leave 25750 minutes.

Nightly binaries build + nightly tests = 3050 minutes.

25750 / 3050 = 8.44. So we could do both 8 more times for additional CI runs for any reason. And that is with pretty good safety margin.

**Skip test_tensorexpr**
On s390x, pytorch is built without llvm.
Even if it would be built with llvm, llvm currently doesn't support used features on s390x and test fails with errors like:
```
JIT session error: Unsupported target machine architecture in ELF object pytorch-jitted-objectbuffer
unknown file: Failure
C++ exception with description "valOrErr INTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/torch/csrc/jit/tensorexpr/llvm_jit.h":34, please report a bug to PyTorch. Unexpected failure in LLVM JIT: Failed to materialize symbols: { (main, { func }) }
```
**Disable cpp/static_runtime_test on s390x**

Quantization is not fully supported on s390x in pytorch yet.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @peterbell10

AlekseiNikiforovIBM committed:
s390x: run periodic tests on s390x

AlekseiNikiforovIBM commented:
@pytorchbot rebase

ezyang commented:
this one's for dev infra, I fear...

AlekseiNikiforovIBM commented:
@malfet @huydhn , could you please take a look at this PR? And if it gets merged, where could be seen the results of periodic tests of new workflow?

malfet commented on test/test_ops_gradients.py:
```diff
@@ -30,6 +31,9 @@ class TestBwdGradients(TestGradients):
     # Tests that gradients are computed correctly
     @_gradcheck_ops(op_db + hop_db + custom_op_db)
     def test_fn_grad(self, device, dtype, op):
+        if platform.machine() == "s390x":
```
Why not use IS_S390 here?

AlekseiNikiforovIBM replied:
I'll double check if this is possible and will update it if it's possible.

malfet commented on test/test_autograd.py:
```diff
@@ -8116,6 +8119,7 @@ def backward(ctx, x):
         c = Func.apply(a)
         self.assertEqual(repr(c), "tensor([2.], grad_fn=<FuncBackward>)")
 
+    @unittest.skipIf(IS_S390X, "Fails on s390x CI")
```
You've introduced skipIfS390, why not use it?

AlekseiNikiforovIBM replied:
skipIfS390X is used only in one place, only because I haven't found how to use IS_S390X there. Everywhere else IS_* is used, so I'm trying to stick to more used solution for skipping.

malfet commented on test/run_test.py:
```diff
@@ -186,6 +188,219 @@ def __contains__(self, item):
+    "mobile/test_lite_script_module",
+    "mobile/test_lite_script_type",
+    "mobile/test_quantize_fx_lite_script_module",
+    "onnx/debug_embed_params",
```
Looks like S390 does not support any ONNX or say inductor test. In that case, please use globs.

AlekseiNikiforovIBM replied:
There are a lot of currently skipped inductor tests, but it's still less than half of them. I'll double check onnx cases, since recently some onnx-related dependencies were added, and if still all fails, I'll use glob.

malfet commented on .github/workflows/s390x-periodic.yml:
```diff
@@ -0,0 +1,71 @@
+  schedule:
+    # We have several schedules so jobs can check github.event.schedule to activate only for a fraction of the runs.
+    # Also run less frequently on weekends.
+    - cron: 29 8 * * *  # about 1:29am PDT, for mem leak check and rerun disabled tests
```
I think I've asked this question in another PR: can you please add some napkin math calculations on how much capacity such testing will require vs what is available right now.
And if capacity is low, please use separate tag other that ciflow/periodic

AlekseiNikiforovIBM replied:
Build took 60 minutes
Tests took: 150, 110, 65, 55, 115, 85, 50, 70, 105, 110 minutes (split into 10 shards)

60 + 150 + 110 + 65 + 55 + 115 + 85 + 50 + 70 + 105 + 110 = 975 minutes used. Let's double it. It would be 1950 minutes.

We have 20 machines * 24 hours = 20 * 24 * 60 = 20 * 1440 = 28800 minutes

We currently run 5 nightly binaries builds, each on average 90 minutes build, 15 minutes test, 5 minutes upload, 110 minutes total for each, 550 minutes total. Doubling would be 1100 minutes.

That leaves 28800 - 1100 = 27700 minutes total. Periodic tests would use will leave 25750 minutes.

Nightly binaries build + nightly tests = 3050 minutes.

25750 / 3050 = 8.44. So we could do both 8 more times for additional CI runs for any reason. And that is with pretty good safety margin.

malfet commented on .github/workflows/_linux-test-s390x.yml:
```diff
@@ -0,0 +1,268 @@
+name: linux-test-s390x
```
Why linux-test.yml can not be reused here?

AlekseiNikiforovIBM replied:
I will try reusing it.

malfet reviewed (commented):
Please provide some capacity estimate and explain why new -build/-test yml file are needed

AlekseiNikiforovIBM commented:
build took 60 minutes
tests took: 150, 110, 65, 55, 115, 85, 50, 70, 105, 110 minutes (split into 10 shards)
There are currently 20 runners, used for everything: nightly binaries, CI runs, and possibly these tests as well.

As for _linux-build.yml vs _linux-build-s390x.yml, the main difference between them is that s390x version doesn't have "docker-image" output.

https://github.com/pytorch/pytorch/blob/main/.github/workflows/_linux-build.yml#L133-L151

It's skipped in s390x since those calculations involve AWS access, which s390x runners don't have. Here's the access:
https://github.com/pytorch/test-infra/blob/main/.github/actions/calculate-docker-image/action.yml#L94-L102

In addition to that, since s390x runners don't have AWS access, data between build and test workers is passed via GHA. I think it was mentioned that using AWS is preferrable. Would that be problem in that case?

I was not yet successful in skipping this step on s390x and not failing build. I can't find logs now, but I can try reproducing it again.

As for _linux-test.yml vs _linux-test-s390x.yml, there are not so many differences, I will try porting all important parts to original version.

huydhn commented:
@pytorchbot rebase

huydhn commented:
Answer to the capacity question https://github.com/pytorch/pytorch/pull/125399#issuecomment-2345746062

huydhn commented on .ci/docker/requirements-ci.txt:
```diff
@@ -298,7 +298,7 @@ pytest-cpp==2.3.0
 #Pinned versions: 2.3.0
 #test that import:
 
-z3-solver==4.12.2.0
```
Note from the OH: we upgrade z3-solver because it has a fix for s390x (@AlekseiNikiforovIBM  to provide more details)

huydhn commented on .ci/pytorch/build.sh:
```diff
@@ -345,11 +347,11 @@ else
     CUSTOM_OP_BUILD="${CUSTOM_TEST_ARTIFACT_BUILD_DIR}/custom-op-build"
     CUSTOM_OP_TEST="$PWD/test/custom_operator"
     python --version
-    SITE_PACKAGES="$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')"
```
Note from OH: why could we just continue tio use distuils here?

huydhn commented on test/run_test.py:
```diff
@@ -185,6 +187,181 @@ def __contains__(self, item):
     "distributed/_tensor/test_attention",
 ]
 
+S390X_BLOCKLIST = [
```
From the OH: The blocklist here it too long, maybe we could provide an allow list here instead, i.e. like running only test X, Y, Z

malfet commented on test/run_test.py:
```diff
@@ -185,6 +187,181 @@ def __contains__(self, item):
 ]
 
+S390X_BLOCKLIST = [
+    # following tests fail for various reasons
```
If S390 fails on so many tests, it would perhaps be better to just have a short op-in list in test.sh (similar to how it's currently done for linux-aarch64)

malfet replied:
For example see https://github.com/pytorch/pytorch/blob/68272ab5967f448ed6d2986039a0ef0ddf0e1b37/test/run_test.py#L1421

malfet commented on .github/workflows/_linux-test.yml:
```diff
@@ -278,19 +297,34 @@ jobs:
+          if [[ ${BUILD_ENVIRONMENT} == *"s390x"* ]]; then
+            # building grpcio requires system openssl on s390x
+            # protobuf is required for onnx
+            docker exec -t "${container_name}" sh -c "dnf install -y protobuf-devel protobuf-c-devel protobuf-lite-devel ; env GRPC_PYTHON_BUILD_SYSTEM_OPENSSL=True pip3 install grpcio==1.65.4"
+
+            # ml_dtypes isn't built for s390x, and last release fails to build
+            # upstream already patched build system to build with newer numpy, apply those changes
+            # build and install ml_dtypes
+            docker exec -t "${container_name}" sh -c "dnf install -y wget patch && cd ~ && git clone https://github.com/jax-ml/ml_dtypes && cd ml_dtypes && git checkout v0.4.0 && git submodule update --init --recursive && wget https://github.com/jax-ml/ml_dtypes/commit/b969f76914d6b30676721bc92bf0f6021a0d1321.patch && wget https://github.com/jax-ml/ml_dtypes/commit/d4e6d035ecda073eab8bcf60f4eef57 ...
```
This should be part of the Docker build

huydhn commented on .github/workflows/_linux-test.yml:
```diff
@@ -278,19 +297,34 @@ jobs:
-          docker exec -t "${container_name}" sh -c "pip install $(echo dist/*.whl)[opt-einsum] && ${TEST_COMMAND}"
+
+          if [[ ${BUILD_ENVIRONMENT} == *"s390x"* ]]; then
+            # building grpcio requires system openssl on s390x
```
Notes from OH: We need to move this to the Docker image for s390x

huydhn replied:
Notes: from OH: we should have a ciflow tag to run s390x on demand, then go with periodic

malfet reviewed (changes requested):
Let's start with the opt in S390 testing, and make sure that for a few weeks it's fine

AlekseiNikiforovIBM commented:
Separate label was mentioned to run these tests. How should it be named?

huydhn commented:
> Separate label was mentioned to run these tests. How should it be named?

There is already one called `ciflow/s390` in https://github.com/pytorch/pytorch/blob/main/.github/pytorch-probot.yml#L19, It's best if you could use that

malfet commented on .github/workflows/s390x-periodic.yml:
```diff
@@ -0,0 +1,68 @@
+    tags:
+      - ciflow/s390/*
+    branches:
+      - release/*
```
Please add a periodic dispatch here (or remove periodic from it's name)

malfet commented on .ci/pytorch/common_utils.sh:
```diff
@@ -82,13 +82,22 @@ function pip_install_whl() {
+    pip install --progress-bar off "$@" || pip install --progress-bar off "$@" || pip install --progress-bar off "$@" ||\
+    pip install "$@" || pip install "$@" || pip install "$@"
+  else
+    pip3 install --progress-bar off "$@" || pip3 install --progress-bar off "$@" || pip3 install --progress-bar off "$@" ||\
+    pip3 install "$@" || pip3 install "$@" || pip3 install "$@"
```
Please submit separate PR (or dev infra can take care of it), as change is not specific to S390, we can use pip3 everywhere.

malfet replied:
https://github.com/pytorch/pytorch/pull/140185

malfet commented on test/functorch/test_ops.py:
```diff
@@ -1041,6 +1042,10 @@ def fn(inp, *args, **kwargs):
                 xfail(
                     "unbind_copy"
                 ),  # Batching rule not implemented for aten::unbind_copy.int.
+                decorate("linalg.tensorsolve", decorator=skipIfS390X),
+                decorate("nn.functional.max_pool1d", decorator=skipIfS390X),
+                decorate("nn.functional.max_unpool2d", decorator=skipIfS390X),
+                decorate("nn.functional.multilabel_margin_loss", decorator=skipIfS390X),
```
This sounds fine to me, but why not submit it first, and then add automation later

malfet commented on test/test_ops_gradients.py:
```diff
@@ -30,6 +31,9 @@ class TestBwdGradients(TestGradients):
     # Tests that gradients are computed correctly
     @_gradcheck_ops(op_db + hop_db + custom_op_db)
     def test_fn_grad(self, device, dtype, op):
+        if IS_S390X:
+            self.skipTest("Skipped! Investigate issue on s390x!")
+            return
```
Please use the decorator

malfet commented on test/test_tensorboard.py:
```diff
@@ -410,6 +411,7 @@ def test_video(self):
         summary.video('dummy', np.random.rand(20, 7, 1, 8, 8))
 
     @unittest.skipIf(IS_MACOS, "Skipping on mac, see https://github.com/pytorch/pytorch/pull/109349 ")
+    @unittest.skipIf(IS_S390X, "Fails on s390x")
```
Why not use `skipIfS390X` here?

malfet reviewed (changes requested):
Please break it down into several PRs:
 - One that moves from pip to pip3 (DevInfra will take care of that)
 - Land changes that introduces skips but no CI (and either use `skipIfS390` or delete the wrapper)
 - Add change that adds CI

AlekseiNikiforovIBM commented:
@pytorchbot rebase

AlekseiNikiforovIBM committed:
Update tests state for s390x

AlekseiNikiforovIBM commented:
It looks ready. @malfet @huydhn , could you please take one more look at this PR?

huydhn commented:
@pytorchbot rebase

AlekseiNikiforovIBM committed:
Satisfy linter for workflow file

AlekseiNikiforovIBM commented:
@pytorchbot rebase

AlekseiNikiforovIBM commented:
@pytorchbot rebase

AlekseiNikiforovIBM commented:
@huydhn, I've fixed new lint issues. And I rebased it a couple of times. Other commits broke some tests a couple of times, but it's now fixed in main branch.

malfet committed:
Merge branch 'main' into s390x_ci_periodic_tests

huydhn commented on .github/workflows/s390x-periodic.yml:
```diff
@@ -0,0 +1,77 @@
+    - cron: 29 8 * * *  # about 1:29am PDT, for mem leak check and rerun disabled tests
+  push:
+    tags:
+      - ciflow/periodic/*
```
we should use a separate ciflow tag for this one, i.e. `ciflow/s390x-periodic`

malfet commented:
@pytorchbot merge -r

malfet commented:
@pytorchbot merge -f "Rebase failed, but let's see if it's mergeable"
